provider/pkg/provider/team_environment_perm.go

```diff
@@ -0,0 +1,184 @@
+		Team:         permId.Team,
+		Environment:  permId.Environment,
+		Permission:   *permission,
+	}
+
+	properties, err := plugin.MarshalProperties(inputs.ToPropertyMap(), plugin.MarshalOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to marshal inputs to properties: %w", err)
+	}
+	return &pulumirpc.ReadResponse{
+		Id:         req.Id,
+		Properties: properties,
+		Inputs:     properties,
+	}, nil
+}
+
+func (tp *PulumiServiceTeamEnvironmentPermissionResource) Create(req *pulumirpc.CreateRequest) (*pulumirpc.CreateResponse, error) {
+	ctx := context.Background()
+	var inputs TeamEnvironmentPermissionInput
+	err := serde.FromProperties(req.GetProperties(), structTagKey, &inputs)
+	if err != nil {
+		return nil, err
+	}
+	request := pulumiapi.CreateTeamEnvironmentPermissionRequest{
+		TeamEnvironmentPermissionRequest: pulumiapi.TeamEnvironmentPermissionRequest{
+			Organization: inputs.Organization,
+			Team:         inputs.Team,
+			Environment:  inputs.Environment,
+		},
+		Permission: inputs.Permission,
+	}
+
+	err = tp.client.AddEnvironmentPermission(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+
+	environmentPermissionId := teamEnvironmentPermissionId{
+		Organization: inputs.Organization,
+		Team:         inputs.Team,
+		Environment:  inputs.Environment,
+		Permission:   inputs.Permission,
+	}
+
+	return &pulumirpc.CreateResponse{
+		Id:         environmentPermissionId.String(),
+		Properties: req.GetProperties(),
+	}, nil
+}
+
+func (tp *PulumiServiceTeamEnvironmentPermissionResource) Delete(req *pulumirpc.DeleteRequest) (*pbempty.Empty, error) {
+	ctx := context.Background()
+	var inputs TeamEnvironmentPermissionInput
+	err := serde.FromProperties(req.GetProperties(), structTagKey, &inputs)
+	if err != nil {
+		return nil, err
+	}
+	request := pulumiapi.TeamEnvironmentPermissionRequest{
+		Organization: inputs.Organization,
+		Team:         inputs.Team,
+		Environment:  inputs.Environment,
+	}
+	err = tp.client.RemoveEnvironmentPermission(ctx, request)
+	if err != nil {
+		return nil, err
+	}
+	return &pbempty.Empty{}, nil
+}
+
+func (tp *PulumiServiceTeamEnvironmentPermissionResource) Diff(req *pulumirpc.DiffRequest) (*pulumirpc.DiffResponse, error) {
+	changedKeys, err := serde.DiffOldsAndNews(req)
+	if err != nil {
+		return nil, err
+	}
+	changes := pulumirpc.DiffResponse_DIFF_NONE
+	if len(changedKeys) > 0 {
+		changes = pulumirpc.DiffResponse_DIFF_SOME
+	}
+	return &pulumirpc.DiffResponse{
+		Changes:  changes,
+		Replaces: changedKeys,
+	}, nil
+}
+
+// Update does nothing because we always replace on changes, never an update
+func (tp *PulumiServiceTeamEnvironmentPermissionResource) Update(_ *pulumirpc.UpdateRequest) (*pulumirpc.UpdateResponse, error) {
+	return nil, fmt.Errorf("unexpected call to update, expected create to be called instead")
+}
+
+type teamEnvironmentPermissionId struct {
+	Organization string
+	Team         string
+	Environment  string
+	Permission   string
```

Comment:
We should remove the `Permission` here. What the ID represents here is really a reference to the team whose permission is being managed, not the value of that permission.

Reply:
Makes sense, removed permission from ID

+ edited the description `unopened` -> `environment definition`